src/serac/physics/utilities/finite_element_state.hpp

```diff
@@ -241,23 +247,182 @@ class FiniteElementState {
+  detail::MaybeOwningPointer<mfem::ParGridFunction> gf_;
+  mfem::HypreParVector                              true_vec_;
+  std::string                                       name_ = "";
+};
+
+/**
+ * @brief A shim class for describing the interface of something that can be synced
+ */
+class SyncableData {
+public:
+  virtual ~SyncableData() = default;
+  virtual void sync()     = 0;
+};
+
+/**
+ * @brief Stores instances of user-defined type for each quadrature point in a mesh
+ * @tparam T The type of the per-qpt data
+ * @pre T must be default-constructible (TODO: Do we want to allow non-default constructible types?)
+ * @pre T must be trivially copyable (due to the use of memcpy for type punning)
+ */
+template <typename T>
+class QuadratureData : public SyncableData {
+public:
+  /**
+   * @brief Constructs using a mesh and polynomial order
+   * @param[in] mesh The mesh for which quadrature-point data should be stored
+   * @param[in] p The polynomial order of the associated finite elements
+   */
+  QuadratureData(mfem::Mesh& mesh, const int p, const bool alloc = true);
+
+  QuadratureData(mfem::QuadratureFunction& qfunc)
+      : qspace_(qfunc.GetSpace()), qfunc_(&qfunc), data_(static_cast<std::size_t>(qfunc.Size() / stride_))
+  {
+    const double* qfunc_ptr = detail::retrieve(qfunc_).GetData();
+    int           j         = 0;
+    T*            data_ptr  = data_.data();
+    for (int i = 0; i < detail::retrieve(qfunc_).Size(); i += stride_) {
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(data_ptr + j, qfunc_ptr + i, sizeof(T));
+      j++;
+    }
+  }
+
+  /**
+   * @brief Retrieves the data for a given quadrature point
+   * @param[in] element_idx The index of the desired element within the mesh
+   * @param[in] q_idx The index of the desired quadrature point within the element
+   */
+  T& operator()(const int element_idx, const int q_idx);
+
+  /**
+   * @brief Assigns an item to each quadrature point
+   * @param[in] item The item to assign
+   */
+  QuadratureData& operator=(const T& item);
+
+  /**
+   * @brief Iterator to the data for the first quadrature point
+   */
+  auto begin() { return data_.begin(); }
+  /// @overload
+  auto begin() const { return data_.begin(); }
+  /**
+   * @brief Iterator to one element past the data for the last quadrature point
+   */
+  auto end() { return data_.end(); }
+  /// @overload
+  auto end() const { return data_.end(); }
+
+  mfem::QuadratureFunction& QFunc() { return detail::retrieve(qfunc_); }
+
+  /**
+   * @brief Synchronizes data from the stored vector<T> to the raw double*
+   * array used by the underlying mfem::QuadratureFunction
+   *
+   * Used for saving to a file - MFEMSidreDataCollection
+   * (and by extension mfem::DataCollection's interface) only allow for
+   * quadrature-point-specific data via mfem::QuadratureFunction, so this logic
+   * is needed to glue together a generic array of data with that class
+   */
```

Comment:
This is a good comment, since the name "sync()" alone doesn't convey which way the data is being copied (i.e. `T* -> double*` vs `double* -> T*`), but it might be better placed on the base class, to ensure that classes inheriting from `SyncableData` follow the same convention.

Reply:
Good suggestion, I've added a comment (containing similar info) to the `SyncableData` Doxygen annotation.